packages/aws-cdk-lib/aws-efs/lib/efs-file-system.ts

```diff
@@ -301,10 +318,75 @@ abstract class FileSystemBase extends Resource implements IFileSystem {
    * @param actions The actions to grant
    */
   public grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant {
-    return iam.Grant.addToPrincipal({
+    return iam.Grant.addToPrincipalOrResource({
+      grantee: grantee,
+      actions: actions,
+      resourceArns: [this.fileSystemArn],
+      resource: this,
+    });
+  }
```

Comment:
Change the grant function from iam.Grant.addToPrincipal to iam.Grant.addToPrincipalOrResource in grant method, like a same resource that has resource policy.

Examples
- https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-s3/lib/bucket.ts#L968-L988
- https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-sns/lib/topic-base.ts#L134C1-L141
- https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-kms/lib/key.ts#L164-L196